forge-ai/src/main/java/forge/ai/ComputerUtil.java

```diff
@@ -1165,42 +1177,244 @@ public static boolean castPermanentInMain1(final Player ai, final SpellAbility s
+                final String validPlayer = sta.getParam("ValidPlayer");
+                if (validPlayer == null || !validPlayer.contains("You")) return true;
+            }
+
+            // 16+ cards match `S:Mode\$ CantBlockBy.*ValidAttacker\$[^|]*You`
+            // Bower Passage etc
+            if (mode.contains(StaticAbilityMode.CantBlockBy)) {
+                final String validAttacker = sta.getParam("ValidAttacker");
+                if (validAttacker != null && validAttacker.contains("You")) return true;
+            }
+
+            // 9+ cards match `S:Mode\$ IgnoreLegendRule`
+            // allow cloning your commander with Mirror Gallery; Myriad etc
+            if (mode.contains(StaticAbilityMode.IgnoreLegendRule)) {
+                return true;
             }
-        }
 
+            // 8+ cards match `S:Mode\$ CantBeActivated.*Activator\$[^|]*(?!You)`
+            // Drana and Linvala; City of Solitude
+            if (mode.contains(StaticAbilityMode.CantBeActivated)) {
+                final String activator = sta.getParam("Activator");
+                if ("Opponent".equals(activator) || "Player.NonActive".equals(activator)) return true;
+            }
+
+            // 11+ card match `S:Mode\$ CantPreventDamage.*(?!Card.Self)`
+            // Everlasting Torment etc
+            if (mode.contains(StaticAbilityMode.CantPreventDamage)) {
+                final String validSource = sta.getParam("ValidSource");
+                if (!"Card.Self".equals(validSource)) return true;
+            }
+
+            // 4+ cards match `S:Mode\$ CanAttackDefender.*ValidCard\$[^|]*You`
+            // Guardians of Oboro etc
+            if (mode.contains(StaticAbilityMode.CanAttackDefender)) {
+                if (validCard.contains("You")) return true;
+            }
+
+            // 4+ cards match `S:Mode\$ NoCleanupDamage`
+            // Patient Zero etc
+            if (mode.contains(StaticAbilityMode.NoCleanupDamage)) {
+                if (!"Card.Self".equals(validCard)) return true;
+            }
+
+            // 3+ cards match `S:Mode\$ ActivateAbilityAsIfHaste.*ValidCard\$[^|]*You`
+            // Thousand-Year Elixir etc
+            if (mode.contains(StaticAbilityMode.ActivateAbilityAsIfHaste)) {
+                if (validCard.contains("You")) return true;
+            }
+
+            // 3+ cards match `S:Mode\$ IgnoreHexproof`
+            // Glaring Spotlight etc
+            if (mode.contains(StaticAbilityMode.IgnoreHexproof)) {
+                return true;
+            }
+
+            // Everlasting Torment
+            if (mode.contains(StaticAbilityMode.WitherDamage)) {
+                return true;
+            }
+        }  // static abilities
+
+        // Activated abilities
+        for (SpellAbility nma : cardState.getNonManaAbilities()) {
+            // 1710+ cards match `A:AB\$.*Cost\$[^|]*Sac`
+            // sac as combat trick
+            final String cost = nma.getParam("Cost");
```

Comment:
Wouldn't it be better to check the real Cost Object?

You might want to check if the Sac Cost also Contains the Tap Cost (unless the creature would have haste)